File: typescript/src/secret.ts
import { SampError } from "./error.js";

export class Seed {
  private constructor(private readonly bytes: Uint8Array) {}

  static fromBytes(b: Uint8Array): Seed {
    if (b.length !== 32) throw new SampError(`seed must be 32 bytes, got ${b.length}`);
    return new Seed(b);
  }

  // WHY: every caller is an audit point — grep `exposeSecret` to enumerate them.
  static exposeSecret(s: Seed): Uint8Array {
    return s.bytes;
  }

  toString(): string {
    return "Seed([REDACTED])";
  }
}

export class ViewScalar {
  private constructor(private readonly bytes: Uint8Array) {}

  static fromBytes(b: Uint8Array): ViewScalar {
    if (b.length !== 32) throw new SampError(`view_scalar must be 32 bytes, got ${b.length}`);
    return new ViewScalar(b);
  }

  static exposeSecret(v: ViewScalar): Uint8Array {
    return v.bytes;
  }

  toString(): string {
    return "ViewScalar([REDACTED])";
  }
}

export class ContentKey {
  private constructor(private readonly bytes: Uint8Array) {}

  static fromBytes(b: Uint8Array): ContentKey {
    if (b.length !== 32) throw new SampError(`content_key must be 32 bytes, got ${b.length}`);
    return new ContentKey(b);
  }

  static exposeSecret(c: ContentKey): Uint8Array {
    return c.bytes;
  }

  toString(): string {
    return "ContentKey([REDACTED])";
  }
}

File: go/secret.go
package samp

type Seed struct{ b [32]byte }

func SeedFromBytes(b [32]byte) Seed { return Seed{b} }

// ExposeSecret returns the raw 32 bytes. Every caller of this method is an
// audit point: it is the only way to reach seed material.
func (s Seed) ExposeSecret() [32]byte { return s.b }

func (s Seed) String() string { return "Seed([REDACTED])" }

type ContentKey struct{ b [32]byte }

func ContentKeyFromBytes(b [32]byte) ContentKey { return ContentKey{b} }
func (c ContentKey) ExposeSecret() [32]byte     { return c.b }
func (c ContentKey) String() string             { return "ContentKey([REDACTED])" }

type ViewScalar struct{ b [32]byte }

func ViewScalarFromBytes(b [32]byte) ViewScalar { return ViewScalar{b} }
func (v ViewScalar) ExposeSecret() [32]byte     { return v.b }
func (v ViewScalar) String() string             { return "ViewScalar([REDACTED])" }
